db/system_keyspace.cc

```diff
@@ -1684,6 +1684,30 @@ future<> system_keyspace::remove_endpoint(gms::inet_address ep) {
     co_await force_blocking_flush(PEERS);
 }
 
+future<> system_keyspace::store_quarantined_hosts(std::unordered_set<locator::host_id> host_ids) {
+    static sstring req = format("INSERT into {}.{} (key, host_id) VALUES ('{}', ?) USING TTL 2592000", NAME, QUARANTINED_HOSTS, QUARANTINED_HOSTS);
```

Comment:
I wonder how the TTL mechanism will interact with node restarts.

It seems that we may keep hosts quarantined indefinitely given the right pattern of restarts.
In fact, if I understand correctly, the recommended pattern of restarts - i.e. rolling restarts - will cause hosts to be kept in this set indefinitely.

Indeed, suppose that the quarantine of a host expires at a certain node. The node restarts and `load_quarantined_hosts` doesn't return that host. However, other nodes which still haven't restarted (we're doing a rolling restart) are gossiping it. We notice we don't have the host, so we add it with a refreshed TTL. And so on...

Reply:
Well, this table should be maintained by Raft when enabled.


To solve the problem you identified, we could insert the host id together with its removal timestamp, and skip the re-insertion if 30 days have passed ("manual TTL"). For Raft it's unnecessary since only the topology change coordinator touches it.

Reply:
> Well, this table should be maintained by Raft when enabled.
> 

Raft topology change does not _need_ this table. It has its own view of topology and nodes that left are a natural part of it. Everything we add now to gossiper and topology change code is just a burden for the Raft project.

Reply:
> Well, this table should be maintained by Raft when enabled.
> 
> > Raft topology change does not need this table. It has its own view of topology and nodes that left are a natural part of it. Everything we add now to gossiper and topology change code is just a burden for the Raft project.

I think it's worth paying this price for the interim until we have a better solution with raft.

> To solve the problem you identified, we could insert the host id together with its removal timestamp, and skip the re-insertion if 30 days have passed ("manual TTL"). For Raft it's unnecessary since only the topology change coordinator touches it.

We shouldn't reinsert anything if a quarantined node restarts while it is already quarantined since it would be listed in storage_service::_quarantined_hosts (that is loaded from system.quarantined_hosts).

If it restarts after the 30 days quarantine passed, then I believe it will look like a bootstrapping node and it may get its tokens back.  This would clearly be an operator error.  Of course, we can extend the quarantine TTL to a longer period to lower that risk.

Reply:
> > Well, this table should be maintained by Raft when enabled.
> > > Raft topology change does not need this table. It has its own view of topology and nodes that left are a natural part of it. Everything we add now to gossiper and topology change code is just a burden for the Raft project.
> 
> I think it's worth paying this price for the interim until we have a better solution with raft.
> 

It is always worth paying the price for the one who does not pay it :) Anyway it should be for project management to decide. With raft this table will be unused, but a recovery mode (if we will ever have one for raft topology. Going from raft to gossiper is not as easy as for schema changes),  it will have to populate it. OTOH if you in recovery maintaining quarantined nodes may be the least of our problems.

Reply:
> > > Well, this table should be maintained by Raft when enabled.
> > 
> > 
> > Raft topology change does not _need_ this table. It has its own view of topology and nodes that left are a natural part of it. Everything we add now to gossiper and topology change code is just a burden for the Raft project.
> 
> Maybe, but we don't want tables maintained by gossip while under Raft control.

Than do not add one. Or it will be simply empty and unused with raft.

Reply:
The goal of the series is to simplify the Raft group 0 integration by getting rid of the last barrier to unifying the raft group 0 host id with the general scylla host id. So raft stands to benefit from this change.

Reply:
This is a false premise. All the raft code is already written and committed. If we want to we can just rename raft_id into host_id and host_id into legacy_host_id and forget about it. It will be used in gossiper mode only. The reason the series exists is because of "id phobia". Somehow getting rid of one id will make live easier despite having hundreds if not thousands of other ids. But this comes with a price - all this code here which will be abandoned in raft mode.

Reply:
> But this comes with a price - all this code here which will be abandoned in raft mode.

Raft would be optional for quite a while.  When disabled, we'd still use the quarantined hosts mechanism, on top of the unique host_id, for safety.

Reply:
And that's is something for project management to decide. Do we want to continue extending gossiper with features and code that will not be reusable in raft mode, but will have to be maintained for a long time to fix problems that was there forever. Or do we want to make some features raft only and are ready to wait for them.
Related question is: when raft topology is available, but optional does adding a feature to it requires adding one to gossiper mode as well?

Reply:
> All the raft code is already written and committed.

But we can remove a lot of the code related to maintaining RAFT_IDs in gossiper and on disk if we migrate to HOST_ID:
https://github.com/kbr-/scylla/tree/host-id-as-raft-id
It will improve the maintainability of the codebase a lot.

>  If we want to we can just rename raft_id into host_id and host_id into legacy_host_id and forget about it.

RAFT_IDs of other nodes may be missing during upgrade. All the code that uses HOST_ID assumes it is present for every node at all times. There's a lot complexity in reasoning about RAFT_IDs. It would be good to get rid of it.

>  It will be used in gossiper mode only.

Which we will have to keep supporting for quite a while.

> But this comes with a price - all this code here which will be abandoned in raft mode.

I think it's fine, we'll remove it once Raft topology is GA and mandatory.
For me the pros outweigh the cons.

> Related question is: when raft topology is available, but optional does adding a feature to it requires adding one to gossiper mode as well?

We could restrict **new** features to Raft-only mode.